### PR TITLE
Adds reportback install hook

### LINKED ISSUE #6014: Add run nid to current signups and reportbacks
after (or at the same time as) #5981 
all signups and reportbacks should have run id added to the col
currently they all default to 0
image

refs #5995 #5993

### PR DESCRIPTION
**Pull request by itsjoekent**
#### What's this PR do?

Adds reportback install hook which 
#### Where should the reviewer start?

https://media.giphy.com/media/fDO2Nk0ImzvvW/giphy.gif
#### How should this be manually tested?

Use a query like this (replace the #'s with your own)

```
SELECT n.nid, rb.rbid
FROM node n
INNER JOIN node run ON run.nid = 1903 
INNER JOIN dosomething_reportback rb 
    ON rb.nid = n.nid 
    AND (rb.run_nid = 0 OR rb.run_nid IS NULL)
AND n.nid = 1485
```

To test for the ones which have a date range, add this line `AND from_unixtime(rb.created) BETWEEN '" . $run->start_date . "' AND '" . $run->end_date . "'` and get the run/start dates from this query:

```
SELECT DISTINCT c.entity_id as run_nid, d.field_run_date_value as start_date, d.field_run_date_value2 as end_date
                      FROM field_data_field_campaigns c
                      INNER JOIN field_data_field_run_date d ON d.entity_id = c.entity_id
                      WHERE c.bundle = 'campaign_run'
                      AND c.field_campaigns_target_id = 1485
                      ORDER BY d.field_run_date_value2 ASC
```

After running the update that query shouldnt return anything 
#### Any background context you want to provide?

I mean, I think this update is pretty simple? (In all seriousness, see the million comments left in the code)
#### What are the relevant tickets?

Fixes #6014 
#### Questions

I think we're gonna need to block off time again to quickly test this @angaither @mikefantini ? (also probably back the database up incase something fucks up?)

### CODE REVIEW

**Comment by itsjoekent** [comment]
@angaither is there anything sena needs to do before we merge this? should i be targeting a new branch instead of dev?

**Comment by angaither** [comment]
@deadlybutter yeah, it'll probably be better to have this be a feature branch we can deploy on it's own

File: lib/modules/dosomething/dosomething_reportback/dosomething_reportback.install [hunk 1]
```diff
@@ -887,3 +887,69 @@ function dosomething_reportback_update_7033() {
...
+                    FROM node n
+                    WHERE n.type = 'campaign'");
+
+  // Loop through every NID
+  foreach ($nids as $nid) {
+
+    // Get the runs for associated for this campaign & the timing
+    $runs = db_query("SELECT DISTINCT c.entity_id as run_nid, d.field_run_date_value as start_date, d.field_run_date_value2 as end_date
+                      FROM field_data_field_campaigns c
+                      INNER JOIN field_data_field_run_date d ON d.entity_id = c.entity_id
```
Comment (angaither) [inline lib/modules/dosomething/dosomething_reportback/dosomething_reportback.install]: i think you would need to join on language here as well

Comment (itsjoekent) [inline lib/modules/dosomething/dosomething_reportback/dosomething_reportback.install]: to prevent duplicates?

File: lib/modules/dosomething/dosomething_reportback/dosomething_reportback.install [hunk 2]
```diff
@@ -887,3 +887,69 @@ function dosomething_reportback_update_7033() {
...
+                                AND (rb.run_nid = 0 OR rb.run_nid IS NULL)
+                             AND n.nid = " . $nid->nid);
+
+        // For reach reportback returned
+        foreach($results as $result) {
+          // Load the reportback and change its run nid to the corresponding
+          // one we've now discovered.
+          $rb = reportback_load($result->rbid);
+          $rb->run_nid = $run->run_nid;
+          entity_save('reportback', $rb);
```
Comment (angaither) [inline lib/modules/dosomething/dosomething_reportback/dosomething_reportback.install]: you could abstract this into a helper function

**Comment by angaither** [comment]
If you don't want this to be a feature branch, you could turn it into a script so we can deploy it whenever we are ready

also, it will make the deploy faster!

**Comment by itsjoekent** [comment]
ah, script idea sounds good.

File: lib/modules/dosomething/dosomething_reportback/dosomething_reportback.install (hunk 1 above)
Comment (angaither) [inline lib/modules/dosomething/dosomething_reportback/dosomething_reportback.install]: no, to get the date that matches the language of the run

**Comment by itsjoekent** [comment]
@angaither ok i changed a bunch of files, this should probably get re-reviewed

File: lib/modules/dosomething/dosomething_reportback/dosomething_reportback.drush.inc [hunk 3]
```diff
@@ -1,8 +1,92 @@
...
+  foreach ($nids as $nid) {
+
+    // Get the runs for associated for this campaign & the timing
+    $runs = db_query("SELECT DISTINCT c.entity_id as run_nid, d.field_run_date_value as start_date, d.field_run_date_value2 as end_date
+                      FROM field_data_field_campaigns c
+                      INNER JOIN field_data_field_run_date d ON d.entity_id = c.entity_id AND d.language = 'en'
+                      WHERE c.bundle = 'campaign_run'
+                      AND c.language = 'en'
+                      AND c.field_campaigns_target_id =" . $nid->nid ."
+                      ORDER BY d.field_run_date_value2 ASC");
```
Comment (angaither) [inline lib/modules/dosomething/dosomething_reportback/dosomething_reportback.drush.inc:44]: the syntax looks strange here, might need to alternate ' "

Comment (itsjoekent) [inline lib/modules/dosomething/dosomething_reportback/dosomething_reportback.drush.inc:44]: do you mean switch all of the double quotes to single? sorry not following

Comment (angaither) [inline lib/modules/dosomething/dosomething_reportback/dosomething_reportback.drush.inc:44]: alternate between single and double quotes

File: lib/modules/dosomething/dosomething_signup/dosomething_signup.module [hunk 4]
```diff
@@ -217,6 +217,7 @@ function dosomething_signup_create($nid, $uid = NULL, $source = NULL, $timestamp
 
   $campaign = Campaign::get($nid);
   $run = dosomething_helpers_get_current_campaign_run_for_user($nid, $user, $campaign);
+  $run->nid = 0;
```
Comment (angaither) [inline lib/modules/dosomething/dosomething_signup/dosomething_signup.module]: why is this added?

File: lib/modules/dosomething/dosomething_signup/includes/dosomething_signup.inc [hunk 5]
```diff
@@ -95,8 +95,8 @@ class SignupEntityController extends EntityAPIController {
     // If the entity uid doesnt belong to current user:
     if ($entity->uid != $user->uid) {
       // And current user can't edit any reportback:
-      if (!user_access('edit any signup')) {
-        watchdog('dosomething_signup', "Attempted uid override for @entity by User @uid", 
+      if (!user_access('edit any signup') && !drupal_is_cli()) {
```
Comment (angaither) [inline lib/modules/dosomething/dosomething_signup/includes/dosomething_signup.inc:98]: does this work? I would think it would fail out above when it checks the entity uid and the user? 
this should probably be it's own line, like [in the global module](https://github.com/DoSomething/phoenix/blob/dev/lib/modules/dosomething/dosomething_global/dosomething_global.module#L24)

File: lib/modules/dosomething/dosomething_signup/dosomething_signup.module (hunk 4 above)
Comment (itsjoekent) [inline lib/modules/dosomething/dosomething_signup/dosomething_signup.module]: leftover from testing (quick hack to make it not assign run nid upon signup), thanks for spotting that

File: lib/modules/dosomething/dosomething_signup/includes/dosomething_signup.inc (hunk 5 above)
Comment (itsjoekent) [inline lib/modules/dosomething/dosomething_signup/includes/dosomething_signup.inc:98]: yep this does work, @sergii-tkachenko did the same thing for reportback saves here https://github.com/DoSomething/phoenix/blob/dev/lib/modules/dosomething/dosomething_reportback/includes/ReportbackController.php#L106

File: lib/modules/dosomething/dosomething_reportback/dosomething_reportback.drush.inc (hunk 3 above)
Comment (itsjoekent) [inline lib/modules/dosomething/dosomething_reportback/dosomething_reportback.drush.inc:44]: Yeah on line 40/41 its single quotes, but the query is surrounded by double. i'm missing something?

**Comment by angaither** [comment]
:+1: 
@deadlybutter please merge, deploy and test on thor with the campaigns that have multiple runs but didn't archive data
